libkineto/src/RoctracerActivityApi.cpp

```diff
@@ -75,6 +80,8 @@ int RoctracerActivityApi::processActivities(
   clock_gettime(CLOCK_REALTIME, &t00);
 
   const timestamp_t toffset = (timespec_to_ns(t0) >> 1) + (timespec_to_ns(t00) >> 1) - timespec_to_ns(t1);
+  startTime = (startTime * 1000) - toffset;
```

Comment:
Could you add a small comment about the time conversion? Also, we should probably move the timestamps in Kineto to TSC clock in the future.

Reply:
I will.  Generally you would use CLOCK_MONOTONIC.  It always moves forward so it does not invert the order of any two events.  Used by e.g.: roctracer, all my tools, and viztracer.

Cupti is pushing out CLOCK_REALTIME timestamps.  So you will see events invert whenever an NTP update happens.  So you will see a kernel execute before it was queued. :o

By TSC I think you effectively mean CLOCK_MONOTONIC_RAW.  That will drift over time (months of uptime) as it makes no adjustment.  This will become a problem when doing multi-node profiling.  You will have to implement a clock synchronization protocol that runs at collection time.  CLOCK_MONOTONIC on NTP systems may be close enough for government work.

And while we are on the topic: pytorch and kineto are using usec timestamps which are course enough to cause collisions.  This leads to things like: 1) events having zero duration and 2) calling order (stack frames) being ambiguous.

Oh last bit: add a libkineto::getTimestamp() function.  In that above code I am using CLOCK_REALTIME empirically to convert the timestamps.  I should really be asking kineto what the current timestamp is and adjusting to that.  Current code will be broken if you convert away from the current timestamp domain.